### PR TITLE
perf(sync): keyset pagination for /sync/* catalog RPCs (#103)

### LINKED ISSUE #103: Switch sync RPCs to keyset pagination (drop offset-linear scaling)
## Background

Full-catalog `campfire sync` recently started hitting a Postgres statement timeout at ~29k/30k objects:

```
API error: 500 - {"error":"Failed to fetch objects","details":"canceling statement due to statement timeout"}
```

The three `/api/v1/sync/*` endpoints all use OFFSET-based pagination:
- `get_objects_for_sync` — `supabase/schemas/functions.sql:133` (ORDER BY `o.object_id`)
- `get_spectra_for_sync` — `supabase/schemas/functions.sql:289` (ORDER BY `s.spectrum_id`)
- `get_photometry_for_sync` — `supabase/schemas/functions.sql:381` (ORDER BY `op.id`)

OFFSET forces Postgres to walk `offset + limit` ordered rows per page, so deep pages grow linearly in cost. `get_objects_for_sync` compounds this: the MATERIALIZED `matched` CTE runs three aggregate CTEs (`member_targets_agg`, `spectra_agg`, `lists_agg`) on top of that scan. The per-page floor is high enough that by page ~29 we tip past the service_role statement timeout.

**Hotfix shipped in #TBD-hotfix-PR:** raised `statement_timeout` on those three RPCs to `120s` via `supabase/migrations/20260421160000_bump_sync_rpc_timeout.sql`. This unblocks current users but just pushes the cliff out; it will reappear as the catalog grows.

## Proposed solution: keyset pagination

Each `/sync/*` endpoint scans **forward only** — perfect fit for keyset. Replace OFFSET with a cursor on the existing ordering column:

| RPC                       | Cursor column                | Index backing it             |
|---------------------------|------------------------------|------------------------------|
| `get_objects_for_sync`    | `objects.object_id` (text)   | `objects_object_id_key` UNIQUE btree |
| `get_spectra_for_sync`    | `spectra.spectrum_id` (text) | `idx_spectra_spectrum_id` UNIQUE btree (`supabase/schemas/indexes.sql:158`) |
| `get_photometry_for_sync` | `object_photometry.id` (int) | PK btree                     |

All three columns are already UNIQUE with a btree, so no new indexes required. Each page becomes O(log N + limit) instead of O(offset + limit).

### Query shape

```sql
-- before
WHERE o.programs && p_program_slugs
  AND o.is_active = true
  AND (p_updated_since IS NULL OR o.updated_at > p_updated_since)
ORDER BY o.object_id
LIMIT p_limit OFFSET p_offset

-- after
WHERE o.programs && p_program_slugs
  AND o.is_active = true
  AND (p_updated_since IS NULL OR o.updated_at > p_updated_since)
  AND (p_after_object_id IS NULL OR o.object_id > p_after_object_id)
ORDER BY o.object_id
LIMIT p_limit
```

## Scope

**SQL (`supabase/schemas/functions.sql` + migration):**
- `get_objects_for_sync`: add `p_after_object_id text DEFAULT NULL`, add predicate, keep `p_offset` for backward compat.
- `get_spectra_for_sync`: add `p_after_spectrum_id text DEFAULT NULL`.
- `get_photometry_for_sync`: add `p_after_id integer DEFAULT NULL`.
- New `GRANT EXECUTE` lines for the new signatures.

**Web (`web/app/api/v1/sync/*/route.ts`):**
- Accept `?after=<cursor>` query param; pass through to the RPC call (`supabase.rpc(..., { p_after_object_id: ... })`).

**Python (`python/campfire/api/client.py`):**
- `_paginate_sync_endpoint` (line 227): drop the `offset` counter; track the last item's key (passed in as `cursor_key`, e.g. `\"object_id\"` or `\"spectrum_id\"`) and send it as `after` on the next request. Stop when the page is empty.
- `fetch_all_objects` / `fetch_all_spectra`: pass `cursor_key`.
- `fetch_all_photometry` (line 332): same refactor, `cursor_key=\"id\"`.

**Cleanup (follow-on):**
- Drop `SET statement_timeout = '120s'` from the three RPCs in `functions.sql` and in a new migration once new clients are the norm.

## Edge cases / design notes

- **Uniqueness.** Cursor columns are all UNIQUE — no need for compound `(sort_col, id)` tiebreaker predicates. Leave a comment noting that any future change to the ORDER BY column must preserve this or switch to row-value comparison.
- **Progress reporting.** Keyset can't answer \"N of M\" cheaply. First-page call still requests `include_counts=true`, which returns `total`; subsequent pages just increment `fetched` by page length. `tqdm` already handles `total` being set once up front — see `_make_progress` in `python/campfire/sync.py:29`.
- **Backwards compat.** Keep `p_offset` as a no-op path when `p_after_*` is NULL. Old Python clients keep working under the 120s timeout until they're updated; new clients skip the timeout entirely.
- **Non-sync call sites unaffected.** UI list pages use different RPCs (`get_filtered_object_ids`, `get_filtered_spectra_paginated`) that genuinely need random-access paging — those keep OFFSET.
- **Tests.** No existing tests exercise `_paginate_sync_endpoint` directly (checked `python/tests/test_store.py`); the refactor should add at least one that verifies cursor propagation (mocking `APISession.get`).

## Acceptance criteria

- [ ] Full `campfire sync --full` at 100k+ objects completes without timeout.
- [ ] Per-page latency is flat (not growing with `fetched`).
- [ ] Old Python clients (offset-based) still work against the updated server.
- [ ] `SET statement_timeout = '120s'` can be removed from the three RPCs in a follow-up migration with no regression.

## Related

- Hotfix: `supabase/migrations/20260421160000_bump_sync_rpc_timeout.sql`
- Previous scaling work: commit `ac4007f` (\"Scale sync RPCs + misc polish\") — MATERIALIZE'd the `matched` CTE and gated count CTEs.

### PR DESCRIPTION
**Pull request by hollisakins**
Closes #103.

## Problem
OFFSET pagination is linear in offset: each `/sync/*` page scans+orders `offset + limit` rows, so deep `campfire sync --full` pages grow in cost until they tip past the 120s service_role statement timeout (the hotfix in `20260423160000_bump_sync_rpc_timeout.sql` just pushed the cliff out). This reappears as the catalog grows.

## Fix
All four `/sync/*` RPCs scan forward only — a perfect fit for keyset. Each gains a `p_after_*` cursor on its unique, btree-indexed ordering column, making each page `O(log N + limit)`:

| RPC | Cursor | Index |
|---|---|---|
| `get_objects_for_sync` | `objects.object_id` | `objects_object_id_key` |
| `get_spectra_for_sync` | `spectra.spectrum_id` | `idx_spectra_spectrum_id` |
| `get_photometry_for_sync` | `object_photometry.id` | pkey |
| `get_storage_objects_for_sync` | `storage_objects.id` | pkey |

## Deviations from the (April-2026) issue text
1. **Four RPCs, not three** — the issue predates `get_storage_objects_for_sync` (epic #210 storage unification), which is the *largest* sync catalog and also OFFSET-based.
2. **Storage cursors on `id`, not `storage_key`** — `storage_key` is only UNIQUE as `(backend,bucket,storage_key)`, so keyset on it could skip/dup rows. Sync order is irrelevant (the client upserts by key). Its RPC is also **restructured**: the old shared `scoped` CTE was referenced 3× (materialized → full scan every page, defeating keyset), so `matched` now restates the scope inline against the base table and seeks the PK index; `scoped` (trimmed to `updated_at`) feeds only the gated count CTEs.
3. **Photometry got `p_include_counts` gating** — it was the only sync RPC counting on every page — and folds into the shared paginator.

## Subtle correctness fix
The client cursors on the **last** returned element (`items[-1][cursor_key]`), so each final `jsonb_agg(...)` now has an explicit `ORDER BY <cursor_col>`. The `matched` CTE's ordering does **not** survive into the aggregate (objects has LEFT JOINs after `matched`, and PG doesn't guarantee aggregate input order). Without this, keyset would silently skip/dup rows.

## Backward compatibility
`p_offset` and `SET statement_timeout = '120s'` are **retained** on all four RPCs so old offset-based clients keep working. Both can be dropped in a follow-up migration once keyset clients are the norm.

## Changes
- **SQL**: `supabase/schemas/functions.sql` + migration `20260711202253_keyset_sync_pagination.sql`. (The migration was hand-trimmed of a spurious matview drop+recreate that `supabase db diff` emits for pre-existing, unrelated matview drift — see #228.)
- **Web**: the four `web/app/api/v1/sync/*/route.ts` accept `?after=<cursor>`.
- **Python**: `_paginate_sync_endpoint(path, cursor_key, ...)` is keyset-based; `fetch_all_photometry` folds into it (duplicate offset loop removed). +3 tests.

## Validation (local Supabase, 89 obj / 138 spec / 138 storage seed)
- Keyset walk == offset walk == full-scan enumeration for objects/spectra/storage, verified with the last-element cursor exactly as the client does.
- Confirmed `Index Scan using objects_object_id_key` with `object_id > cursor` + early `LIMIT` stop.
- Count-gating, empty-photometry, past-end-cursor edge cases pass; storage scope-leak SQL tests still green.
- 452 pytest pass, 1 skipped; `npm run build` green.

No pipeline changelog entry (touches `web/`, `python/`, `supabase/` only).

Generated with Claude Code

### CODE REVIEW

File: python/campfire/api/client.py [hunk 1]
```diff
@@ -436,28 +454,16 @@ def fetch_all_photometry(
...
+        """Fetch all photometry records via the /sync/photometry endpoint.
+
+        Keyset cursor is the integer ``id`` (object_photometry PK). The second
+        return element is 0 here (the endpoint carries no accessible count) and
+        is ignored by the caller; kept for signature parity with the other
+        ``fetch_all_*`` streams.
+        """
+        return self._paginate_sync_endpoint(
+            "/sync/photometry", "id", updated_since, on_page_complete,
+        )
```
[the remark this answers is not in the document]
Comment (hollisakins) [reply python/campfire/api/client.py:473]: @claude can you fix?
